### PR TITLE
170 fix shap error cat features

### LINKED ISSUE #170: EarlyStoppingShapRFECV with LightGBM - Error after all categorical features have been removed
The following error occurs once all categorical features have been removed:

AttributeError: 'TreeEnsemble' object has no attribute 'values'

This is due to a known issue with SHAP and LightGBM (https://github.com/slundberg/shap/issues/480). Check on line 98 of shap_helpers.py usually deals with this issue by calling SHAP explainer differently when categoricals are present. However when all categorical features have been dropped the error still presents itself. Quick solution is to set a flag in the beginning if any categorical features are present and then use that flag instead of the check on line 98. Would be good to understand why this happens though.

**Environment:**

 - probatus 1.8.5
 - Shap 0.38.1
 - LightGBM 3.2.1 
 - python 3.8.2
 - OS: linux

**To Reproduce**

```python
from probatus import feature_elimination
from sklearn.model_selection import StratifiedKFold
import lightgbm as lgb
import pandas as pd

X = pd.DataFrame(
    [
        [1,2,3,4,5,101,102,103,104,105],
        [-1,-2,2,-5,-7,1,2,5,-1,3],
        ["a", "b"] * 5, # noisy categorical will dropped first
    ]
).transpose()
X[2] = X[2].astype("category")
X[1] = X[1].astype("float")
X[0] = X[0].astype("float")
y = [0] * 5 + [1] * 5

model = lgb.LGBMClassifier()

n_iter = 2

n_folds = 3

for i in range(n_iter):
    skf = StratifiedKFold(n_folds, shuffle=True, random_state=42)
    shap_elimination = feature_elimination.EarlyStoppingShapRFECV(
        clf=model,
        step=1 / (n_iter + 1),
        cv=skf,
        scoring="accuracy",
        eval_metric="logloss",
        early_stopping_rounds=5,
    )
    report = shap_elimination.fit_compute(
        X,
        y,
    )
```

**Error traceback**

If applicable please provide full traceback of the error.

```
Traceback (most recent call last):
  File "/home/luke/.virtualenvs/af/lib/python3.8/site-packages/joblib/externals/loky/process_executor.py", line 431, in _process_worker
    r = call_item()
  File "/home/luke/.virtualenvs/af/lib/python3.8/site-packages/joblib/externals/loky/process_executor.py", line 285, in __call__
    return self.fn(*self.args, **self.kwargs)
  File "/home/luke/.virtualenvs/af/lib/python3.8/site-packages/joblib/_parallel_backends.py", line 595, in __call__
    return self.func(*args, **kwargs)
  File "/home/luke/.virtualenvs/af/lib/python3.8/site-packages/joblib/parallel.py", line 262, in __call__
    return [func(*args, **kwargs)
  File "/home/luke/.virtualenvs/af/lib/python3.8/site-packages/joblib/parallel.py", line 262, in <listcomp>
    return [func(*args, **kwargs)
  File "/home/luke/.virtualenvs/af/lib/python3.8/site-packages/probatus/feature_elimination/feature_elimination.py", line 1079, in _get_feature_shap_values_per_fold
    shap_values = shap_calc(
  File "/home/luke/.virtualenvs/af/lib/python3.8/site-packages/probatus/utils/shap_helpers.py", line 108, in shap_calc
    explainer = Explainer(model, masker=mask, **shap_kwargs)
  File "/home/luke/.virtualenvs/af/lib/python3.8/site-packages/shap/explainers/_explainer.py", line 161, in __init__
    explainers.Tree.__init__(self, self.model, self.masker, link=self.link, feature_names=self.feature_names, **kwargs)
  File "/home/luke/.virtualenvs/af/lib/python3.8/site-packages/shap/explainers/_tree.py", line 177, in __init__
    self.expected_value = self.model.predict(self.data).mean(0)
  File "/home/luke/.virtualenvs/af/lib/python3.8/site-packages/shap/explainers/_tree.py", line 1091, in predict
    tree_limit = self.values.shape[0]
AttributeError: 'TreeEnsemble' object has no attribute 'values'
```

**Expected behavior**

Should run fine even once all categorical features are dropped.

### PR DESCRIPTION
**Pull request by ClaudioSalvatoreArcidiacono**
Fixes #170 

With this change a workaround for the issue https://github.com/slundberg/shap/issues/480 is proposed.

The workaround consists in not passing any masking data when `feature_perturbation` is explicitly set to `tree_path_dependent` in the `fit_compute` method or when the dataset contains categorical features. This will force shap explainer to use the `tree_path_dependent` strategy (instead of `interventional`), which supports categorical features explanation and should not raise the exception:

```
AttributeError: 'TreeEnsemble' object has no attribute 'values'
```

A very nice explanation of the differences between `tree_path_dependent` and `interventional` can be found here https://github.com/slundberg/shap/issues/1098#issuecomment-599622022.

### CODE REVIEW

File: probatus/utils/shap_helpers.py [hunk 1]
```diff
@@ -85,31 +85,30 @@ def shap_calc(
...
+        # https://github.com/slundberg/shap/issues/480
+        if shap_kwargs.get("feature_perturbation") == "tree_path_dependent" or X.select_dtypes("category").shape[1] > 0:
+            # Calculate Shap values.
             explainer = Explainer(model, **shap_kwargs)
         else:
+            # Create the background data,required for non tree based models.
+            # A single datapoint can passed as mask
+            # (https://github.com/slundberg/shap/issues/955#issuecomment-569837201)
+            if X.shape[0] < sample_size:
+                sample_size = int(np.ceil(X.shape[0] * 0.2))
```
Comment (gverbock) [inline probatus/utils/shap_helpers.py:100]: Should we pass the 0.2 as a parameter?

File: probatus/utils/shap_helpers.py [hunk 2]
```diff
@@ -85,31 +85,30 @@ def shap_calc(
...
-            if verbose > 0:
-                warnings.warn(
-                    "Using tree_dependent feature_perturbation (in shap) without background"
-                    " data for tree-based model with categorical features."
-                )
+        # For tree explainers, do not pass masker when feature_perturbation is
+        # tree_path_dependent, or when X contains categorical features
+        # related to issue:
+        # https://github.com/slundberg/shap/issues/480
+        if shap_kwargs.get("feature_perturbation") == "tree_path_dependent" or X.select_dtypes("category").shape[1] > 0:
```
Comment (gverbock) [inline probatus/utils/shap_helpers.py:92]: Wasn't it that mask is only "required" for logistic regression @anilkumarpanda?

Comment (ClaudioSalvatoreArcidiacono) [reply probatus/utils/shap_helpers.py:92]: Mask is required for non Tree-based models.
For Tree-based models you can have two types of `feature_perturbation`: `tree_path_dependent` and `interventional`. The former does not require a background dataset (aka mask), the latter does require it.
It is explained in the shap documentation:
https://shap.readthedocs.io/en/latest/generated/shap.explainers.Tree.html#shap.explainers.Tree.__init__
(see the `feature_perturbation` parameter)

File: probatus/utils/shap_helpers.py (hunk 1 above)
Comment (ClaudioSalvatoreArcidiacono) [reply probatus/utils/shap_helpers.py:100]: This change is out of scope for this issue.

File: probatus/utils/shap_helpers.py (hunk 2 above)
Comment (anilkumarpanda) [reply probatus/utils/shap_helpers.py:92]: True,have a mask is required for non tree based models.

**Pull request merged**